### ISSUE TITLE
Make incremental compile faster on constant changes

### ISSUE THREAD

**Issue by breskeby** [gradle/gradle#6482]
We use the rocker template engine for generating ui and the gradle rocker plugin by @etiennestuder.

In our app we see that changing a rocker template file always triggers a full recompilation of the app source even though that should not be required. 

After consulting with @melix I've learned that @oehme lately worked on incremental compilation. @oehme would be great if you can have a look.

### Expected Behavior
- A rocker template change does not result in a full recompilation

### Current Behavior
- A rocker template change results in a full recompilation

### Context
The GE team is affected by this issue.

### Steps to Reproduce (for bugs)

An Example build is attached:
[rocker-incremental-compile.zip](https://github.com/gradle/gradle/files/2314094/rocker-incremental-compile.zip)

1. Run `./gradlew compileJava` on example project
2. Change content in `src/rocker/Greetz.rocker.html` (e.g change the greeting message)
3. Run `./gradlew compileJava -d`. This will result in a full recompile. the debug output provides the following information: 

```
Full recompilation is required because 'Greetz.java' was changed
```

I've tested the behaviour with 4.9, 4.10-rc-2 and latest gradle master (Gradle 5.0-20180823103201+0000)

This behaviour seemed to indicate a constant change in the class. When checking the changes in Greetz.java (generated by the rocker plugin in `src/generated/rocker/Greetz.java` seems that only private static fields have changed. So a full recompilation shouldn't be required.

**Comment by oehme** [comment]
Does that class have public constants? If so this is expected, any change to a file with constant will result in a full recompile. We don't parse sources, so we don't know what actually changed.

Workaround would be to compile those templates in their own compile task and put the result on the main classpath. That way the full recompile would be limited to the templates only.

**Comment by breskeby** [comment]
thanks for checking. Indeed it has: ```

    static public final com.fizzed.rocker.ContentType CONTENT_TYPE = com.fizzed.rocker.ContentType.HTML;
    static public final String TEMPLATE_NAME = "Greetz.rocker.html";
    static public final String TEMPLATE_PACKAGE_NAME = "";
    static public final String HEADER_HASH = "973482035";
    static public final String[] ARGUMENT_NAMES = { "message" };

```

I'll check if there's a way to tweak the template generation.

**Comment by breskeby** [comment]
Some further findings:

- incremental compilation works when going back to gradle 3.4. 
- having all rocker templates in one folder leads to a high chance of recompilation as all compiled templates contain a public constant string with the shared package name. 
- using templates with same parameter names also raises the chance for a bigger recompile as a constant `HEADER_HASH` is based on those parameter names. 

I assume with comments from @oehme above we know about this change in behaviour in Gradle and that it was done on purpose?

**Comment by breskeby** [comment]
As I understand incremental compilation was broken with Gradle < 3.4 and 3.4 fixed the handling of inlined constants. 
Though, the fact that we do a full recompile in the latest gradle version (and do not with 3.4) as soon as we detect a constant at all _is_ another change in the behavior of the incremental compiler. FYI the current behavior does not match up with what we describe in https://blog.gradle.org/incremental-compiler-avoidance

Point 2 & 3 are just about incremental compiler behavior in Gradle 3.4 where in fact the content of the constants do matter (see explanation in https://blog.gradle.org/incremental-compiler-avoidance)

**Comment by oehme** [comment]
Inlined constants can be derived from other constants. 3.4 was broken in that case and 3.5 fixed that.

**Comment by ldaley** [comment]
> Constants are really only meant for things like PI, which is constant forever.

But that is not how they are used in practice, and we are unlikely to be able to change that.

**Comment by ldaley** [comment]
> Workaround would be to compile those templates in their own compile task and put the result on the main classpath

The templates often use classes from the main source so this is more work than it sounds and probably not worth the effort of the extra indirection in the particular case that caused us to raise this.

**Comment by oehme** [comment]
> But that is not how they are used in practice, and we are unlikely to be able to change that.

Case in point - The rocker team is considering such a change. The Android team is also removing the constants from their R classes. And in user code they are often trivial to avoid.

**Comment by ldaley** [comment]
Sure, but we should fully accept the weakness of the “don't use constants” answer.

**Comment by wordlessly** [comment]
Can't you force the compiler to stop inlining public constants (or constants in general) ?

**Comment by oehme** [comment]
No, we can't, because inlining constants is part of the Java language [specification](https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.28). A compiler must do it or it's not a valid Java compiler.

**Comment by wordlessly** [comment]
Convert all public constants into private constants and duplicate them (again as private constants) when used in external classes.

**Comment by oehme** [comment]
We won't change people's code :)

**Comment by wordlessly** [comment]
You already do, by using a compiler :)

Also you could offer annotations for people to tell you to change their code. So instead of removing public constants people could annotate them and you do the rest :)

**Comment by Jazzepi** [comment]
> > But that is not how they are used in practice, and we are unlikely to be able to change that.
> 
> Case in point - The rocker team is considering such a change. The Android team is also removing the constants from their R classes. And in user code they are often trivial to avoid.
> 
> Of course avoiding the cost altogether is the desired goal, hence this issue. But education can unblock some people in the meantime.

I'm glad you guys have an issue open for this. I use public static final String constants in quite a few places. The idea of gutting them and replacing them with methods because Gradle has no idea what the compiler is doing is **just gross.** This is very tail wagging the dog situation. A build system's behavior around incremental compilation should never be a driver for how source code should look.

**Comment by autonomousapps** [comment]
The [documentation](https://docs.gradle.org/current/userguide/java_plugin.html#sec:incremental_compile) on this appears to be misleading. It says:
> Since constants can be inlined, any change to a constant will result in Gradle recompiling all source files. For that reason, you should try to minimize the use of constants in your source code and replace them with static methods where possible.

But shouldn't it read
> Since constants can be inlined, any change to a constant **or a class containing a constant** ...

?

**Comment by hungvietnguyen** [comment]
Or even better:

> Since **public** constants can be inlined, any change to a **public** constant or a class containing a **public** constant... For that reason, you should try to minimize the use of **public** constants...

**Comment by hungvietnguyen** [comment]
Also, the message when running with `--info` says:

> Full recompilation is required because 'SomeClass.java' was changed.

It would be great if it said: 

> Full recompilation is required because 'SomeClass.java' was changed **and it contained a public constant**.

(This confusion came up multiple times at http://issuetracker.google.com/110061530 where developers are unsure why a simple change resulted in full recompile.)

**Comment by wordlessly** [comment]
> Or even better:
> 
> > Since **public** constants can be inlined, any change to a **public** constant or a class containing a **public** constant... For that reason, you should try to minimize the use of **public** constants...

Can be inlined? This is optional?

> No, we can't, because inlining constants is part of the Java language [specification](https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.28). A compiler must do it or it's not a valid Java compiler.

According to @oehme they have to be inlined.

On a different note: can't Gradle "remember" all public constants and check if they have changed since the last compilation run? If not we can skip a full recompile.

**Comment by melix** [comment]
Reopen as improvements are implemented as part of #16495
